// @paraswap/core/build/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DirectContractMethods = exports.DirectContractMethodsWithFeeModel = exports.ContractMethod = exports.ContractMethodV6 = exports.ContractMethodV5 = exports.SwapSide = void 0;
var SwapSide;
(function (SwapSide) {
    SwapSide["BUY"] = "BUY";
    SwapSide["SELL"] = "SELL";
})(SwapSide = exports.SwapSide || (exports.SwapSide = {}));
var ContractMethodV5;
(function (ContractMethodV5) {
    ContractMethodV5["swapOnUniswap"] = "swapOnUniswap";
    ContractMethodV5["buyOnUniswap"] = "buyOnUniswap";
    ContractMethodV5["swapOnUniswapFork"] = "swapOnUniswapFork";
    ContractMethodV5["buyOnUniswapFork"] = "buyOnUniswapFork";
    ContractMethodV5["swapOnUniswapV2Fork"] = "swapOnUniswapV2Fork";
    ContractMethodV5["buyOnUniswapV2Fork"] = "buyOnUniswapV2Fork";
    ContractMethodV5["simpleBuy"] = "simpleBuy";
    ContractMethodV5["simpleSwap"] = "simpleSwap";
    ContractMethodV5["multiSwap"] = "multiSwap";
    ContractMethodV5["megaSwap"] = "megaSwap";
    ContractMethodV5["protectedMultiSwap"] = "protectedMultiSwap";
    ContractMethodV5["protectedMegaSwap"] = "protectedMegaSwap";
    ContractMethodV5["protectedSimpleSwap"] = "protectedSimpleSwap";
    ContractMethodV5["protectedSimpleBuy"] = "protectedSimpleBuy";
    ContractMethodV5["swapOnZeroXv2"] = "swapOnZeroXv2";
    ContractMethodV5["swapOnZeroXv4"] = "swapOnZeroXv4";
    ContractMethodV5["buy"] = "buy";
    ContractMethodV5["directUniV3Swap"] = "directUniV3Swap";
    ContractMethodV5["directUniV3Buy"] = "directUniV3Buy";
    ContractMethodV5["directCurveV1Swap"] = "directCurveV1Swap";
    ContractMethodV5["directCurveV2Swap"] = "directCurveV2Swap";
    ContractMethodV5["directBalancerV2GivenInSwap"] = "directBalancerV2GivenInSwap";
    ContractMethodV5["directBalancerV2GivenOutSwap"] = "directBalancerV2GivenOutSwap";
})(ContractMethodV5 = exports.ContractMethodV5 || (exports.ContractMethodV5 = {}));
var ContractMethodV6;
(function (ContractMethodV6) {
    ContractMethodV6["swapExactAmountIn"] = "swapExactAmountIn";
    ContractMethodV6["swapExactAmountOut"] = "swapExactAmountOut";
    ContractMethodV6["swapExactAmountInOnUniswapV2"] = "swapExactAmountInOnUniswapV2";
    ContractMethodV6["swapExactAmountOutOnUniswapV2"] = "swapExactAmountOutOnUniswapV2";
    ContractMethodV6["swapExactAmountInOnUniswapV3"] = "swapExactAmountInOnUniswapV3";
    ContractMethodV6["swapExactAmountOutOnUniswapV3"] = "swapExactAmountOutOnUniswapV3";
    ContractMethodV6["swapExactAmountInOnBalancerV2"] = "swapExactAmountInOnBalancerV2";
    ContractMethodV6["swapExactAmountOutOnBalancerV2"] = "swapExactAmountOutOnBalancerV2";
    ContractMethodV6["swapExactAmountInOnCurveV1"] = "swapExactAmountInOnCurveV1";
    ContractMethodV6["swapExactAmountInOnCurveV2"] = "swapExactAmountInOnCurveV2";
    ContractMethodV6["swapOnAugustusRFQTryBatchFill"] = "swapOnAugustusRFQTryBatchFill";
    ContractMethodV6["swapExactAmountInOutOnMakerPSM"] = "swapExactAmountInOutOnMakerPSM";
})(ContractMethodV6 = exports.ContractMethodV6 || (exports.ContractMethodV6 = {}));
exports.ContractMethod = { ...ContractMethodV5, ...ContractMethodV6 };
exports.DirectContractMethodsWithFeeModel = [
    ContractMethodV5.directUniV3Swap,
    ContractMethodV5.directUniV3Buy,
    ContractMethodV5.directCurveV1Swap,
    ContractMethodV5.directCurveV2Swap,
    ContractMethodV5.directBalancerV2GivenInSwap,
    ContractMethodV5.directBalancerV2GivenOutSwap,
    ContractMethodV6.swapExactAmountInOnUniswapV2,
    ContractMethodV6.swapExactAmountOutOnUniswapV2,
    ContractMethodV6.swapExactAmountInOnUniswapV3,
    ContractMethodV6.swapExactAmountOutOnUniswapV3,
    ContractMethodV6.swapExactAmountInOnBalancerV2,
    ContractMethodV6.swapExactAmountOutOnBalancerV2,
    ContractMethodV6.swapExactAmountInOnCurveV1,
    ContractMethodV6.swapExactAmountInOnCurveV2,
];
exports.DirectContractMethods = [
    ContractMethodV5.swapOnUniswap,
    ContractMethodV5.buyOnUniswap,
    ContractMethodV5.swapOnUniswapFork,
    ContractMethodV5.buyOnUniswapFork,
    ContractMethodV5.swapOnUniswapV2Fork,
    ContractMethodV5.buyOnUniswapV2Fork,
    ContractMethodV5.swapOnZeroXv2,
    ContractMethodV5.swapOnZeroXv4,
    ContractMethodV6.swapOnAugustusRFQTryBatchFill,
    ContractMethodV6.swapExactAmountInOutOnMakerPSM,
].concat(exports.DirectContractMethodsWithFeeModel);
