// @paraswap/core/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DirectContractMethods = exports.ContractMethodV6 = exports.ContractMethodV5 = exports.ContractMethod = exports.SwapSide = exports.ParaSwapVersion = void 0;
var types_1 = require("./types");
Object.defineProperty(exports, "ParaSwapVersion", { enumerable: true, get: function () { return types_1.ParaSwapVersion; } });
var constants_1 = require("./constants");
Object.defineProperty(exports, "SwapSide", { enumerable: true, get: function () { return constants_1.SwapSide; } });
Object.defineProperty(exports, "ContractMethod", { enumerable: true, get: function () { return constants_1.ContractMethod; } });
Object.defineProperty(exports, "ContractMethodV5", { enumerable: true, get: function () { return constants_1.ContractMethodV5; } });
Object.defineProperty(exports, "ContractMethodV6", { enumerable: true, get: function () { return constants_1.ContractMethodV6; } });
Object.defineProperty(exports, "DirectContractMethods", { enumerable: true, get: function () { return constants_1.DirectContractMethods; } });
