// @protocolink/logics/dist/logics/zeroex-v4/configs.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExchangeProxyAddress = exports.configMap = exports.supportedChainIds = exports.configs = void 0;
const tslib_1 = require("tslib");
const common = tslib_1.__importStar(require("@protocolink/common"));
exports.configs = [
    {
        chainId: common.ChainId.mainnet,
        exchangeProxyAddress: '0xdef1c0ded9bec7f1a1670819833240f027b25eff',
    },
    {
        chainId: common.ChainId.optimism,
        exchangeProxyAddress: '0xdef1abe32c034e558cdd535791643c58a13acc10',
    },
    {
        chainId: common.ChainId.polygon,
        exchangeProxyAddress: '0xdef1c0ded9bec7f1a1670819833240f027b25eff',
    },
    {
        chainId: common.ChainId.base,
        exchangeProxyAddress: '0xdef1c0ded9bec7f1a1670819833240f027b25eff',
    },
    {
        chainId: common.ChainId.arbitrum,
        exchangeProxyAddress: '0xdef1c0ded9bec7f1a1670819833240f027b25eff',
    },
    {
        chainId: common.ChainId.avalanche,
        exchangeProxyAddress: '0xdef1c0ded9bec7f1a1670819833240f027b25eff',
    },
];
_a = exports.configs.reduce((accumulator, config) => {
    accumulator[0].push(config.chainId);
    accumulator[1][config.chainId] = config;
    return accumulator;
}, [[], {}]), exports.supportedChainIds = _a[0], exports.configMap = _a[1];
function getExchangeProxyAddress(chainId) {
    return exports.configMap[chainId].exchangeProxyAddress;
}
exports.getExchangeProxyAddress = getExchangeProxyAddress;
