// @protocolink/logics/dist/logics/openocean-v2/configs.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getApiUrl = exports.getGasPrice = exports.getExchangeAddress = exports.configMap = exports.supportedChainIds = exports.configs = void 0;
const tslib_1 = require("tslib");
const common = tslib_1.__importStar(require("@protocolink/common"));
exports.configs = [
    {
        chainId: common.ChainId.gnosis,
        exchangeAddress: '0x6352a56caadC4F1E25CD6c75970Fa768A3304e64',
        gasPrice: '14',
    },
    {
        chainId: common.ChainId.metis,
        exchangeAddress: '0x6352a56caadC4F1E25CD6c75970Fa768A3304e64',
        gasPrice: '20',
    },
];
_a = exports.configs.reduce((accumulator, config) => {
    accumulator[0].push(config.chainId);
    accumulator[1][config.chainId] = config;
    return accumulator;
}, [[], {}]), exports.supportedChainIds = _a[0], exports.configMap = _a[1];
function getExchangeAddress(chainId) {
    return exports.configMap[chainId].exchangeAddress;
}
exports.getExchangeAddress = getExchangeAddress;
function getGasPrice(chainId) {
    return exports.configMap[chainId].gasPrice;
}
exports.getGasPrice = getGasPrice;
function getApiUrl(chainId) {
    const url = 'https://open-api.openocean.finance/v3';
    return url + `/${chainId}`;
}
exports.getApiUrl = getApiUrl;
