// @protocolink/api/dist/protocols/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zeroexv4 = exports.wagmi = exports.utility = exports.uniswapv3 = exports.syncswap = exports.stargatev2 = exports.stargate = exports.spark = exports.sonne = exports.radiantv2 = exports.permit2 = exports.paraswapv5 = exports.openoceanv2 = exports.morphoblue = exports.magicsea = exports.iolend = exports.compoundv3 = exports.balancerv2 = exports.aavev3 = exports.aavev2 = void 0;
const tslib_1 = require("tslib");
exports.aavev2 = tslib_1.__importStar(require("./aave-v2"));
exports.aavev3 = tslib_1.__importStar(require("./aave-v3"));
exports.balancerv2 = tslib_1.__importStar(require("./balancer-v2"));
exports.compoundv3 = tslib_1.__importStar(require("./compound-v3"));
exports.iolend = tslib_1.__importStar(require("./iolend"));
exports.magicsea = tslib_1.__importStar(require("./magicsea"));
exports.morphoblue = tslib_1.__importStar(require("./morphoblue"));
exports.openoceanv2 = tslib_1.__importStar(require("./openocean-v2"));
exports.paraswapv5 = tslib_1.__importStar(require("./paraswap-v5"));
exports.permit2 = tslib_1.__importStar(require("./permit2"));
exports.radiantv2 = tslib_1.__importStar(require("./radiant-v2"));
exports.sonne = tslib_1.__importStar(require("./sonne"));
exports.spark = tslib_1.__importStar(require("./spark"));
exports.stargate = tslib_1.__importStar(require("./stargate"));
exports.stargatev2 = tslib_1.__importStar(require("./stargate-v2"));
exports.syncswap = tslib_1.__importStar(require("./syncswap"));
exports.uniswapv3 = tslib_1.__importStar(require("./uniswap-v3"));
exports.utility = tslib_1.__importStar(require("./utility"));
exports.wagmi = tslib_1.__importStar(require("./wagmi"));
exports.zeroexv4 = tslib_1.__importStar(require("./zeroex-v4"));
