// @protocolink/logics/dist/logics/magicsea/configs.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBaseTokens = exports.getTokenListUrls = exports.getContractAddress = exports.configMap = exports.supportedChainIds = exports.configs = void 0;
const tslib_1 = require("tslib");
const common = tslib_1.__importStar(require("@protocolink/common"));
const tokens_1 = require("./tokens");
exports.configs = [
    {
        chainId: common.ChainId.iota,
        contract: {
            Factory: '0x349aaAc3a500014981CBA11b64C76c66a6c1e8D0',
            Router: '0x531777F8c35fDe8DA9baB6cC7093A7D14a99D73E',
        },
        tokenListUrls: [
            'https://raw.githubusercontent.com/MagicSea-Finance/tokenlist/main/token.default.json',
            'https://raw.githubusercontent.com/moonpadxyz/tokenlist/main/token.default.json',
        ],
        bases: [common.iotaTokens.wIOTA, tokens_1.iotaTokens['USDC.e'], tokens_1.iotaTokens.USDT],
    },
];
_a = exports.configs.reduce((accumulator, config) => {
    accumulator[0].push(config.chainId);
    accumulator[1][config.chainId] = config;
    return accumulator;
}, [[], {}]), exports.supportedChainIds = _a[0], exports.configMap = _a[1];
function getContractAddress(chainId, name) {
    return exports.configMap[chainId].contract[name];
}
exports.getContractAddress = getContractAddress;
function getTokenListUrls(chainId) {
    return exports.configMap[chainId].tokenListUrls;
}
exports.getTokenListUrls = getTokenListUrls;
function getBaseTokens(chainId) {
    return exports.configMap[chainId].bases;
}
exports.getBaseTokens = getBaseTokens;
