// @protocolink/smart-accounts/dist/configs.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSmartAccount = exports.getSmartAccountIds = exports.getConfig = exports.smartAccountIdMap = exports.smartAccountMap = exports.configMap = exports.supportedChainIds = exports.configs = exports.SmartAccountId = void 0;
const tslib_1 = require("tslib");
const common = tslib_1.__importStar(require("@protocolink/common"));
var SmartAccountId;
(function (SmartAccountId) {
    SmartAccountId["PORTUS"] = "portus";
})(SmartAccountId = exports.SmartAccountId || (exports.SmartAccountId = {}));
exports.configs = [
    {
        chainId: common.ChainId.optimism,
        smartAccounts: [
            {
                id: SmartAccountId.PORTUS,
                executor: '0xdD408b8eFb837EdeF8e6192Ed19f0dbEB7B79383',
            },
        ],
    },
];
_a = exports.configs.reduce((accumulator, config) => {
    accumulator[0].push(config.chainId);
    accumulator[1][config.chainId] = config;
    accumulator[2][config.chainId] = {};
    accumulator[3][config.chainId] = [];
    for (const smartAccount of config.smartAccounts) {
        accumulator[2][config.chainId][smartAccount.id] = smartAccount;
        accumulator[3][config.chainId].push(smartAccount.id);
    }
    return accumulator;
}, [[], {}, {}, {}]), exports.supportedChainIds = _a[0], exports.configMap = _a[1], exports.smartAccountMap = _a[2], exports.smartAccountIdMap = _a[3];
function getConfig(chainId) {
    return exports.configMap[chainId];
}
exports.getConfig = getConfig;
function getSmartAccountIds(chainId) {
    return exports.smartAccountIdMap[chainId];
}
exports.getSmartAccountIds = getSmartAccountIds;
function getSmartAccount(chainId, id) {
    return exports.smartAccountMap[chainId][id];
}
exports.getSmartAccount = getSmartAccount;
