// @protocolink/logics/dist/logics/uniswap-v3/tokens/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.avalancheTokens = exports.arbitrumTokens = exports.baseTokens = exports.polygonTokens = exports.optimismTokens = exports.mainnetTokens = void 0;
const tslib_1 = require("tslib");
const arbitrum_json_1 = tslib_1.__importDefault(require("./data/arbitrum.json"));
const avalanche_json_1 = tslib_1.__importDefault(require("./data/avalanche.json"));
const base_json_1 = tslib_1.__importDefault(require("./data/base.json"));
const common = tslib_1.__importStar(require("@protocolink/common"));
const mainnet_json_1 = tslib_1.__importDefault(require("./data/mainnet.json"));
const optimism_json_1 = tslib_1.__importDefault(require("./data/optimism.json"));
const polygon_json_1 = tslib_1.__importDefault(require("./data/polygon.json"));
const univ3 = tslib_1.__importStar(require("../../../modules/univ3"));
exports.mainnetTokens = {
    ...univ3.toTokenMap(mainnet_json_1.default),
    ...univ3.toTokenMap(common.mainnetTokens),
};
exports.optimismTokens = {
    ...univ3.toTokenMap(optimism_json_1.default),
    ...univ3.toTokenMap(common.optimismTokens),
};
exports.polygonTokens = {
    ...univ3.toTokenMap(polygon_json_1.default),
    ...univ3.toTokenMap(common.polygonTokens),
};
exports.baseTokens = {
    ...univ3.toTokenMap(base_json_1.default),
    ...univ3.toTokenMap(common.baseTokens),
};
exports.arbitrumTokens = {
    ...univ3.toTokenMap(arbitrum_json_1.default),
    ...univ3.toTokenMap(common.arbitrumTokens),
};
exports.avalancheTokens = {
    ...univ3.toTokenMap(avalanche_json_1.default),
    ...univ3.toTokenMap(common.avalancheTokens),
};
