// @protocolink/logics/dist/logics/aave-v2/configs.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContractAddress = exports.configMap = exports.supportedChainIds = exports.configs = exports.protocolId = void 0;
const tslib_1 = require("tslib");
const common = tslib_1.__importStar(require("@protocolink/common"));
exports.protocolId = 'aave-v2';
exports.configs = [
    {
        chainId: common.ChainId.mainnet,
        contract: {
            ProtocolDataProvider: '0x057835Ad21a177dbdd3090bB1CAE03EaCF78Fc6d',
            AaveV2FlashLoanCallback: '0x727c55092C7196d65594A8e4F39ae8dC0cB39173',
        },
    },
    {
        chainId: common.ChainId.polygon,
        contract: {
            ProtocolDataProvider: '0x7551b5D2763519d4e37e8B81929D336De671d46d',
            AaveV2FlashLoanCallback: '0x727c55092C7196d65594A8e4F39ae8dC0cB39173',
        },
    },
    {
        chainId: common.ChainId.avalanche,
        contract: {
            ProtocolDataProvider: '0x65285E9dfab318f57051ab2b139ccCf232945451',
            AaveV2FlashLoanCallback: '0x727c55092C7196d65594A8e4F39ae8dC0cB39173',
        },
    },
];
_a = exports.configs.reduce((accumulator, config) => {
    accumulator[0].push(config.chainId);
    accumulator[1][config.chainId] = config;
    return accumulator;
}, [[], {}]), exports.supportedChainIds = _a[0], exports.configMap = _a[1];
function getContractAddress(chainId, name) {
    return exports.configMap[chainId].contract[name];
}
exports.getContractAddress = getContractAddress;
