// @protocolink/logics/dist/logics/syncswap/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UINT256_MAX = exports.UINT128_MAX = exports.FOUR = exports.THREE = exports.TWO = exports.ONE = exports.ZERO = exports.STABLE_POOL_A = exports.MAX_XP = exports.MAX_LOOP_LIMIT = exports.MAX_FEE = exports.LIQUIDITY_MIN_RESERVE = void 0;
const ethers_1 = require("ethers");
exports.LIQUIDITY_MIN_RESERVE = ethers_1.BigNumber.from(1000);
exports.MAX_FEE = ethers_1.BigNumber.from(100000); // 1e5
exports.MAX_LOOP_LIMIT = 256;
// https://github.com/syncswap/core-contracts/blob/master/contracts/pool/stable/SyncSwapStablePool.sol#L25
exports.MAX_XP = ethers_1.BigNumber.from('3802571709128108338056982581425910818');
exports.STABLE_POOL_A = ethers_1.BigNumber.from(1000);
exports.ZERO = ethers_1.BigNumber.from(0);
exports.ONE = ethers_1.BigNumber.from(1);
exports.TWO = ethers_1.BigNumber.from(2);
exports.THREE = ethers_1.BigNumber.from(3);
exports.FOUR = ethers_1.BigNumber.from(4);
exports.UINT128_MAX = ethers_1.BigNumber.from('0xffffffffffffffffffffffffffffffff');
exports.UINT256_MAX = ethers_1.BigNumber.from('0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff');
