// @protocolink/logics/dist/logics/wagmi/configs.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfig = exports.configMap = exports.supportedChainIds = exports.configs = void 0;
const tslib_1 = require("tslib");
const v3_sdk_1 = require("@uniswap/v3-sdk");
const common = tslib_1.__importStar(require("@protocolink/common"));
const tokens_1 = require("./tokens");
exports.configs = [
    {
        chainId: common.ChainId.iota,
        factoryAddress: '0x01Bd510B2eA106917e711f9a05a42fC162bee2Ac',
        quoter: { address: '0x5C08A6762CAF9ec8a42F249eBC23aAE66097218D', isV2: true },
        swapRouterAddress: '0xaE7b92C8B14E7bdB523408aE0A6fFbf3f589adD9',
        feeAmounts: [v3_sdk_1.FeeAmount.LOWEST - 100, v3_sdk_1.FeeAmount.LOW, v3_sdk_1.FeeAmount.LOW + 1000, v3_sdk_1.FeeAmount.MEDIUM, v3_sdk_1.FeeAmount.HIGH],
        bases: [tokens_1.iotaTokens.wIOTA, tokens_1.iotaTokens.USDT, tokens_1.iotaTokens.WETH],
    },
];
_a = exports.configs.reduce((accumulator, config) => {
    accumulator[0].push(config.chainId);
    accumulator[1][config.chainId] = config;
    return accumulator;
}, [[], {}]), exports.supportedChainIds = _a[0], exports.configMap = _a[1];
function getConfig(chainId) {
    return exports.configMap[chainId];
}
exports.getConfig = getConfig;
