// @protocolink/logics/dist/logics/spark/configs.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContractAddress = exports.configMap = exports.supportedChainIds = exports.configs = exports.protocolId = void 0;
const tslib_1 = require("tslib");
const common = tslib_1.__importStar(require("@protocolink/common"));
exports.protocolId = 'spark';
exports.configs = [
    {
        chainId: common.ChainId.mainnet,
        contract: {
            PoolDataProvider: '0xFc21d6d146E6086B8359705C8b28512a983db0cb',
            SparkFlashLoanCallback: '0x9174a45468d055Cc2Fa18e708E8CeACD46050359',
        },
    },
    {
        chainId: common.ChainId.gnosis,
        contract: {
            PoolDataProvider: '0x2a002054A06546bB5a264D57A81347e23Af91D18',
            SparkFlashLoanCallback: '0x9174a45468d055Cc2Fa18e708E8CeACD46050359',
        },
    },
];
_a = exports.configs.reduce((accumulator, config) => {
    accumulator[0].push(config.chainId);
    accumulator[1][config.chainId] = config;
    return accumulator;
}, [[], {}]), exports.supportedChainIds = _a[0], exports.configMap = _a[1];
function getContractAddress(chainId, name) {
    return exports.configMap[chainId].contract[name];
}
exports.getContractAddress = getContractAddress;
