// @protocolink/smart-accounts/dist/contracts/factories/Executor__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Executor__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        type: 'constructor',
        inputs: [
            {
                name: 'router_',
                type: 'address',
                internalType: 'address',
            },
        ],
        stateMutability: 'nonpayable',
    },
    {
        type: 'function',
        name: 'executeFromAgent',
        inputs: [
            {
                name: 'tos_',
                type: 'address[]',
                internalType: 'address[]',
            },
            {
                name: 'datas_',
                type: 'bytes[]',
                internalType: 'bytes[]',
            },
            {
                name: 'values_',
                type: 'uint256[]',
                internalType: 'uint256[]',
            },
        ],
        outputs: [],
        stateMutability: 'nonpayable',
    },
];
class Executor__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.Executor__factory = Executor__factory;
Executor__factory.abi = _abi;
