// @protocolink/logics/dist/logics/radiant-v2/configs.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContractAddress = exports.configMap = exports.supportedChainIds = exports.configs = exports.protocolId = void 0;
const tslib_1 = require("tslib");
const common = tslib_1.__importStar(require("@protocolink/common"));
exports.protocolId = 'radiant-v2';
exports.configs = [
    {
        chainId: common.ChainId.mainnet,
        contract: {
            ProtocolDataProvider: '0x362f3BB63Cff83bd169aE1793979E9e537993813',
            RadiantV2FlashLoanCallback: '0x6bfCE075A1c4F0fD4067A401dA8f159354e1a916',
        },
    },
    {
        chainId: common.ChainId.bnb,
        contract: {
            ProtocolDataProvider: '0x499e336041202Cd4e55a1979e7511b3211033847',
            RadiantV2FlashLoanCallback: '0x544921763C3D2C9345BA3862eFa72368a12bBd1f',
        },
    },
    {
        chainId: common.ChainId.base,
        contract: {
            ProtocolDataProvider: '0x07d2DC09A1CbDD01e5f6Ca984b060A3Ff31b9EAF',
            RadiantV2FlashLoanCallback: '0x6bfCE075A1c4F0fD4067A401dA8f159354e1a916',
        },
    },
    {
        chainId: common.ChainId.arbitrum,
        contract: {
            ProtocolDataProvider: '0xDd109cb6F2B2aEeBcE01727a31d99E3149aa7e41',
            RadiantV2FlashLoanCallback: '0x544921763C3D2C9345BA3862eFa72368a12bBd1f',
        },
    },
];
_a = exports.configs.reduce((accumulator, config) => {
    accumulator[0].push(config.chainId);
    accumulator[1][config.chainId] = config;
    return accumulator;
}, [[], {}]), exports.supportedChainIds = _a[0], exports.configMap = _a[1];
function getContractAddress(chainId, name) {
    return exports.configMap[chainId].contract[name];
}
exports.getContractAddress = getContractAddress;
